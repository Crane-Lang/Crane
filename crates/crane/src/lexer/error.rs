use serde::{Serialize, Deserialize};
use thiserror::Error;

use crate::ast::Span;

#[derive(Error, Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct LexError {
    pub kind: LexErrorKind,
    pub span: Span,
}

impl Default for LexError {
    fn default() -> Self {
        Self {
            kind: Default::default(),
            span: Span::new(0, 0),
        }
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)
    }
}

#[derive(Debug, Error, PartialEq, Clone, Default, Serialize, Deserialize)]
pub enum LexErrorKind {
    #[default]
    #[error("unknown")]
    Unknown,
}
