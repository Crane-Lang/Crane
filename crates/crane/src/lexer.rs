mod error;

pub mod token;

pub use error::*;

use logos::Logos;

use crate::lexer::token::{Token, TokenKind};

pub struct Lexer<'src> {
    lexer: logos::Lexer<'src, TokenKind>,
}

impl<'src> Lexer<'src> {
    pub fn new(input: &'src str) -> Self {
        Self {
            lexer: TokenKind::lexer(input),
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.lexer.next()?;
        let lexeme = self.lexer.slice();
        let span = self.lexer.span().into();

        Some(
            kind.map(|kind| Token {
                kind,
                lexeme: lexeme.into(),
                span,
            })
            .map_err(|kind| LexError { kind, span }),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(input: &str) -> logos::Lexer<TokenKind> {
        TokenKind::lexer(input)
    }

    fn check(input: &str, token: TokenKind) {
        let mut lexer = TokenKind::lexer(input);

        assert_eq!(lexer.next(), Some(Ok(token)));
        assert_eq!(lexer.slice(), input);
    }

    #[test]
    fn lex_open_paren() {
        check("(", TokenKind::OpenParen)
    }

    #[test]
    fn lex_close_paren() {
        check(")", TokenKind::CloseParen)
    }

    #[test]
    fn lex_open_brace() {
        check("{", TokenKind::OpenBrace)
    }

    #[test]
    fn lex_close_brace() {
        check("}", TokenKind::CloseBrace)
    }

    #[test]
    fn lex_comment() {
        check("// This is a comment.", TokenKind::Comment)
    }

    #[test]
    fn lex_fn_declaration() {
        let mut lex = lex("fn foo() {}");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 0..2);
        assert_eq!(lex.slice(), "fn");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 3..6);
        assert_eq!(lex.slice(), "foo");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenParen)));
        assert_eq!(lex.span(), 6..7);
        assert_eq!(lex.slice(), "(");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseParen)));
        assert_eq!(lex.span(), 7..8);
        assert_eq!(lex.slice(), ")");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenBrace)));
        assert_eq!(lex.span(), 9..10);
        assert_eq!(lex.slice(), "{");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseBrace)));
        assert_eq!(lex.span(), 10..11);
        assert_eq!(lex.slice(), "}");
    }

    #[test]
    fn test_lexer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);

            insta::assert_yaml_snapshot!(lexer.into_iter().collect::<Vec<_>>());
        })
    }
}
