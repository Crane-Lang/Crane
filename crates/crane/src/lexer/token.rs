use logos::Logos;
use smol_str::SmolStr;

use crate::ast::Span;
use crate::lexer::LexErrorKind;

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(error = LexErrorKind)]
pub enum TokenKind {
    /// `(`
    #[token("(")]
    OpenParen,

    /// `)`
    #[token(")")]
    CloseParen,

    /// `{`
    #[token("{")]
    OpenBrace,

    /// `}`
    #[token("}")]
    CloseBrace,

    /// `,`
    #[token(",")]
    Comma,

    /// An identifier.
    #[regex(r"[A-Za-z_]+")]
    Ident,

    /// A string literal.
    #[regex(r#""[^"]*""#)]
    String,

    /// An integer literal.
    #[regex(r#"[\d]+"#)]
    Integer,

    /// A comment.
    #[regex(r"//.*")]
    Comment,

    /// Any sequence of whitespace characters.
    #[regex(r"[ \n]+", logos::skip)]
    Whitespace,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: SmolStr,
    pub span: Span,
}
