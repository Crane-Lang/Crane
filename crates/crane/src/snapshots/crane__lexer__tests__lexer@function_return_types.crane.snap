---
source: crates/crane/src/lexer.rs
expression: "lexer.into_iter().collect::<Vec<_>>()"
input_file: crates/crane/src/snapshot_inputs/function_return_types.crane
---
- Ok:
    kind: Ident
    lexeme: fn
    span:
      start: 0
      end: 2
- Ok:
    kind: Ident
    lexeme: main
    span:
      start: 3
      end: 7
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 7
      end: 8
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 8
      end: 9
- Ok:
    kind: OpenBrace
    lexeme: "{"
    span:
      start: 10
      end: 11
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 16
      end: 23
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 23
      end: 24
- Ok:
    kind: Ident
    lexeme: int_to_string
    span:
      start: 24
      end: 37
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 37
      end: 38
- Ok:
    kind: Ident
    lexeme: add_10
    span:
      start: 38
      end: 44
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 44
      end: 45
- Ok:
    kind: Integer
    lexeme: "5"
    span:
      start: 45
      end: 46
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 46
      end: 47
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 47
      end: 48
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 48
      end: 49
- Ok:
    kind: CloseBrace
    lexeme: "}"
    span:
      start: 50
      end: 51
- Ok:
    kind: Ident
    lexeme: fn
    span:
      start: 53
      end: 55
- Ok:
    kind: Ident
    lexeme: add_10
    span:
      start: 56
      end: 62
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 62
      end: 63
- Ok:
    kind: Ident
    lexeme: n
    span:
      start: 63
      end: 64
- Ok:
    kind: Colon
    lexeme: ":"
    span:
      start: 64
      end: 65
- Ok:
    kind: Ident
    lexeme: Uint64
    span:
      start: 66
      end: 72
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 72
      end: 73
- Ok:
    kind: RightArrow
    lexeme: "->"
    span:
      start: 74
      end: 76
- Ok:
    kind: Ident
    lexeme: Uint64
    span:
      start: 77
      end: 83
- Ok:
    kind: OpenBrace
    lexeme: "{"
    span:
      start: 84
      end: 85
- Ok:
    kind: Ident
    lexeme: int_add
    span:
      start: 90
      end: 97
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 97
      end: 98
- Ok:
    kind: Ident
    lexeme: n
    span:
      start: 98
      end: 99
- Ok:
    kind: Comma
    lexeme: ","
    span:
      start: 99
      end: 100
- Ok:
    kind: Integer
    lexeme: "10"
    span:
      start: 101
      end: 103
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 103
      end: 104
- Ok:
    kind: CloseBrace
    lexeme: "}"
    span:
      start: 105
      end: 106

